### PR TITLE
Magic 2024

### LINKED ISSUE #5: Class adaptation
**Describe your suggestion in full detail below:**
If we want classes to be perk trees, it's time to add them.
`add_perk = <Name>` is syntax for adding perks. Must be in `effect`.

### LINKED ISSUE #1319: Magic GUI
from Magic 2024 planning document: 

GUI Additions:

1. Magic GUI - On the sidebar, like with unions, decisions etc, see basic information about magic. This can be stuff like mana, prowess, progress until next perk, whatever. This can also hold "ability bonuses" and hold secret magic lifestyles. This would also contain a button to open your spellbook
2. Spellbook - This will have a display of spells you’ve unlocked, by magic class. This can also be a front to cast a spell on someone, or yourself. For convenience, the player can store the last 3 characters they casted a spell on, besides themselves. This is basically the same as “Cast Spell” Gui, just accessed differently

### LINKED ISSUE #1338: Magic Duels
@helloruhe suggested special duel mechanics for casters.  let's plan this feature

### PR DESCRIPTION
**Pull request by sumankun**
copied description form Meme PR. Updated to the most current patch

#### Changelog:
- Added Magic Lifestyles.
- Added effects to sorcerer traits
- Changed localisation for magical doctrine parameters
- Close #5

## General TODO:
- [x] Restrict focus selection
- [x] Gender doctrine loc
- [x] Focus buffs / loc
- [ ] Adapt vanilla events
- [ ] Restrict lifestyle selection
- [x] Give magic resistance even if magic usage is secret
- [x] Fix Freezerburn perk that Dione broke
- [ ] Fix chances for army/knight targeting spells (currently one roll for entire army, should be one roll per target)
- [x] Fix spell reflect for new schemes
- [x] Apply Divine Storm to armies and travelers passing through the area
- [x] Banshee, Ghoul, Vampire, Revenant trait modifiers and balancing
- [ ] Epidemic traits balancing
- [ ] Epidemic modifiers
- [x] Fix dodging, hit, and resist TTs
- [x] Fix add or increase burn/frostbite
- [x] Notification when spell is off cooldown
- [x] Notification when spell is finished casting
- [x] Notification toggle in gui
- [ ] Fix court mage + integrate with new court position system
- [x] Add magic modifiers to existing artifacts
- [x] Add magic event options / modifiers to exisitng WC content
- [x] Make sure all magic illnesses / injuries use vanilla treatment system
- [x] Unlearn spells when resetting perks
- [ ] Add monitoring for spells that target lands
- [ ] Make spells that can target countie/duchy also target camps (check dreamwalk and invoke too)
- [x] Update backdraft to monitor characters in area

## Art TODO:
- [x] Magic decision backgrounds
- [ ] Fire ritual icons
- [ ] Life ritual icons
- [ ] Death ritual icons
- [ ] Light ritual icons
- [ ] Shadow ritual icons
- [ ] Fel ritual icons
- [ ] Air ritual icons
- [ ] Arcane ritual icons
- [ ] Earth ritual icons
- [ ] Water ritual icons
- [x] Frostbite decal
- [x] Frostbite trait icon
- [x] Ice Block trait icon
- [x] Voidtouched trait icon
- [x] Frost/snowflake modifier icon
- [x] Ice block static model
- [ ] Master Pyromancer trait icon
- [ ] Master Cryomancer trait icon
- [x] Polymorph trait icon
- [ ] Arcane ritual icons
- [ ] Light ritual icons
- [ ] Abomination 3D static model
- [ ] Banshee appearance
- [ ] Ghoul 3D static model
- [ ] Vampire appearance
- [ ] Vampire trait icon
- [ ] Revenant trait icon
- [x] Frost Fever trait icon
- [x] Crypt Fever trait icon
- [x] Blood Plague trait icon
- [ ] Death ritual icons
- [ ] Water ritual icons
- [ ] Hibernation trait icon
- [ ] Cat form trait icon
- [ ] Bear form trait icon
- [ ] Stag form trait icon
- [ ] Treant form trait icon
- [ ] Cat form static model
- [ ] Bear form static model
- [ ] Stag form static model
- [ ] Treant form static model

## Lifestyles TODO:
- [x] Elemental Fire Lifestyle @helloruhe 
- [x] Elemental Water Lifestyle @robmart 
- [ ] Elemental Air Lifestyle @robmart 
- [ ] Elemental Earth Lifestyle @helloruhe 
- [x] Arcane Lifestyle @robmart 
- [x] Fel Lifestyle @robmart 
- [x] Death Lifestyle @robmart 
- [x] Light Lifestyle @helloruhe 
- [x] Life Lifestyle @helloruhe 
- [x] Shadow Lifestyle @helloruhe 

## Tests:
- [ ] There are no errors in `wc` files in `Documents\Paradox Interactive\Crusader Kings III\logs\error.log`

#### How to test:
- Basically, it's 2nd round of what have been tested in #120 - check that lore character can pick correct focuses in correct lifestyles.
- Ignore perks, they are not ready yet.
- Use `add_lifestyle_xp_all` console command to get exp for all lifestyles

### CODE REVIEW

File: common/traits/wc_lifestyle_traits.txt
Comment (helloruhe) [inline common/traits/wc_lifestyle_traits.txt:1]: why are we removing all of the lifestyle's perk buffs?

File: localization/english/wc_traits_l_english.yml [hunk 1]
```diff
@@ -552,7 +552,33 @@
  TRAIT_FLAG_DESC_goblin_family:0 "[wc_race_family|E]: [wc_goblin_wo_family|E]"
  TRAIT_FLAG_DESC_eredar_family:0 "[wc_race_family|E]: [wc_eredar_wo_family|E]"
 
+ TRAIT_FLAG_DESC_racial_light_magic_resistance:0 "$wc_light_magic_resistance$ #P +5#!"
+ TRAIT_FLAG_DESC_racial_shadow_magic_resistance:0 "$wc_shadow_magic_resistance$ #P +5#!"
+ TRAIT_FLAG_DESC_racial_disorder_magic_resistance:0 "$wc_disorder_magic_resistance$ #P +5#!"
```
Comment (robmart) [inline localization/english/wc_traits_l_english.yml:557]: Duplicate loc

File: localization/english/wc_traits_l_english.yml [hunk 2]
```diff
@@ -552,7 +552,33 @@
...
+ TRAIT_FLAG_DESC_racial_life_magic_resistance:0 "$wc_life_magic_resistance$ #P +5#!"
+ TRAIT_FLAG_DESC_racial_death_magic_resistance:0 "$wc_death_magic_resistance$ #P +5#!"
+ TRAIT_FLAG_DESC_racial_elemental_fire_magic_resistance:0 "$wc_elemental_fire_magic_resistance$ #P +5#!"
+ TRAIT_FLAG_DESC_racial_elemental_water_magic_resistance:0 "$wc_elemental_water_magic_resistance$ #P +5#!"
+ TRAIT_FLAG_DESC_racial_elemental_earth_magic_resistance:0 "$wc_elemental_earth_magic_resistance$ #P +5#!"
+ TRAIT_FLAG_DESC_racial_elemental_air_magic_resistance:0 "$wc_elemental_air_magic_resistance$ #P +5#!" 
+
+ TRAIT_FLAG_DESC_racial_light_magic_vulnerability:0 "$wc_light_magic_resistance$ #N -5#!"
+ TRAIT_FLAG_DESC_racial_shadow_magic_vulnerability:0 "$wc_shadow_magic_resistance$ #N -5#!"
+ TRAIT_FLAG_DESC_racial_disorder_magic_vulnerability:0 "$wc_disorder_magic_resistance$ #N -5#!"
```
Comment (robmart) [inline localization/english/wc_traits_l_english.yml:569]: Duplicate loc

File: common/traits/wc_race_traits.txt [hunk 3]
```diff
@@ -3358,14 +3600,157 @@ creature_lost_one = {
...
+	flag = racial_order_magic_resistance
+
+	flag = undead_class
+	flag = race_trait
+	flag = life_expectancy_3
+	monthly_death_magic_lifestyle_xp_gain_mult = wc_magic_racial_xp_gain_mult_value
+
+	inherit_chance = @race_inherit_chance
+	both_parent_has_trait_inherit_chance = @both_parent_race_inherit_chance
+}
```
Comment (Grekopithikos) [inline common/traits/wc_race_traits.txt]: I don't think vampires should be a race.

In all of the few examples of vampires in WoW, their former race isn't visibly erased like what happens with ghouls and liches.
For example the [San'layn](https://warcraft.wiki.gg/wiki/San%27layn) are still considered to be ([Darkfallen](https://warcraft.wiki.gg/wiki/Darkfallen))elves, [Blood-Thane Lucard](https://warcraft.wiki.gg/wiki/Blood-Thane_Lucard) is still considered to be a vrykul, and [Captain Broketooth](https://warcraft.wiki.gg/wiki/Captain_Broketooth) is considered to be an undead human.

That's why vampires shouldn't be a race of its own but a trait that can be aquired by other races(those that make sense at least).
